src/cryptonote_basic/multisig.cpp

```diff
@@ -0,0 +1,139 @@
+  {
+    CHECK_AND_ASSERT_THROW_MES(view_keys.size() == spend_keys.size(), "Mismatched view/spend key sizes");
+
+    // the multisig spend public key is the sum of all spend public keys
+    spend_pkey = rct::pk2rct(keys.m_account_address.m_spend_public_key);
+    for (const auto &k: spend_keys)
+      rct::addKeys(spend_pkey, spend_pkey, rct::pk2rct(k));
+    multisig_keys.push_back(keys.m_spend_secret_key);
+    spend_skey = rct::sk2rct(keys.m_spend_secret_key);
+  }
+  //-----------------------------------------------------------------
+  void generate_multisig_N1_N(const account_keys &keys, const std::vector<crypto::secret_key> &view_keys, const std::vector<crypto::public_key> &spend_keys, std::vector<crypto::secret_key> &multisig_keys, rct::key &spend_skey, rct::key &spend_pkey)
+  {
+    CHECK_AND_ASSERT_THROW_MES(view_keys.size() == spend_keys.size(), "Mismatched view/spend key sizes");
+
+    spend_pkey = rct::identity();
+    spend_skey = rct::zero();
+
+    // create all our composite private keys
+    for (const auto &k: spend_keys)
+    {
+      rct::keyV data;
+      data.push_back(rct::scalarmultKey(rct::pk2rct(k), rct::sk2rct(keys.m_spend_secret_key)));
+      static const rct::key salt = { {'M', 'u', 'l', 't' , 'i', 's', 'i', 'g' , 0x00, 0x00, 0x00,0x00 , 0x00, 0x00, 0x00,0x00 , 0x00, 0x00, 0x00,0x00 , 0x00, 0x00, 0x00,0x00 , 0x00, 0x00, 0x00,0x00 , 0x00, 0x00, 0x00,0x00  } };
+      data.push_back(salt);
+      rct::key msk = rct::hash_to_scalar(data);
+      multisig_keys.push_back(rct::rct2sk(msk));
+      sc_add(spend_skey.bytes, spend_skey.bytes, msk.bytes);
+    }
+  }
+  //-----------------------------------------------------------------
+  crypto::secret_key generate_multisig_view_secret_key(const crypto::secret_key &skey, const std::vector<crypto::secret_key> &skeys)
+  {
+    crypto::hash hash;
+    crypto::cn_fast_hash(&skey, sizeof(crypto::hash), hash);
+    rct::key view_skey = rct::hash2rct(hash);
+    for (const auto &k: skeys)
+      sc_add(view_skey.bytes, view_skey.bytes, rct::sk2rct(k).bytes);
+    return rct::rct2sk(view_skey);
+  }
+  //-----------------------------------------------------------------
+  crypto::public_key generate_multisig_N1_N_spend_public_key(const std::vector<crypto::public_key> &pkeys)
+  {
+    rct::key spend_public_key = rct::identity();
+    for (const auto &pk: pkeys)
+    {
+      rct::addKeys(spend_public_key, spend_public_key, rct::pk2rct(pk));
+    }
+    return rct::rct2pk(spend_public_key);
+  }
+  //-----------------------------------------------------------------
+  bool generate_multisig_key_image(const account_keys &keys, const crypto::public_key& tx_public_key, size_t real_output_index, cryptonote::keypair& in_ephemeral, crypto::key_image& ki, size_t multisig_key_index)
+  {
+    if (multisig_key_index >= keys.m_multisig_keys.size())
+      return false;
+    if (!cryptonote::generate_key_image_helper(keys, tx_public_key, real_output_index, in_ephemeral, ki))
```

Comment:
AIUI `generate_key_image_helper` returns `false` only when either the tx pubkey or the account spend pubkey is not a proper EC point (i.e. where `ge_frombytes_vartime` fails in `generate_key_derivation` or `derive_public_key`, respectively), so I don't see how the lines below get executed. Am I missing something?